src/sync_device.rs

```diff
@@ -0,0 +1,707 @@
+                    led,
+                    led_vals.contains(led),
+                    led.0
+                )?;
+            }
+        }
+
+        if let Some(supported_snd) = self.supported_sounds() {
+            write!(f, "  Sounds:")?;
+            for snd in supported_snd.iter() {
+                writeln!(f, "    {:?} (index {})", snd, snd.0)?;
+            }
+        }
+
+        // if let Some(rep) = self.rep {
+        //     writeln!(f, "  Repeats: {:?}", rep)?;
+        // }
+
+        let evs = self.supported_events();
+
+        if evs.contains(EventType::FORCEFEEDBACK) {
+            writeln!(f, "  Force Feedback supported")?;
+        }
+
+        if evs.contains(EventType::POWER) {
+            writeln!(f, "  Power supported")?;
+        }
+
+        if evs.contains(EventType::FORCEFEEDBACKSTATUS) {
+            writeln!(f, "  Force Feedback status supported")?;
+        }
+
+        Ok(())
+    }
+}
+
+const fn bus_name(x: u16) -> &'static str {
+    match x {
+        0x1 => "PCI",
+        0x2 => "ISA Plug 'n Play",
+        0x3 => "USB",
+        0x4 => "HIL",
+        0x5 => "Bluetooth",
+        0x6 => "Virtual",
+        0x10 => "ISA",
+        0x11 => "i8042",
+        0x12 => "XTKBD",
+        0x13 => "RS232",
+        0x14 => "Gameport",
+        0x15 => "Parallel Port",
+        0x16 => "Amiga",
+        0x17 => "ADB",
+        0x18 => "I2C",
+        0x19 => "Host",
+        0x1A => "GSC",
+        0x1B => "Atari",
+        0x1C => "SPI",
+        _ => "Unknown",
+    }
+}
+
+#[cfg(feature = "tokio")]
+mod tokio_stream {
+    use super::*;
+
+    use tokio_1 as tokio;
+
+    use crate::nix_err;
+    use futures_core::{ready, Stream};
+    use std::collections::VecDeque;
+    use std::pin::Pin;
+    use std::task::{Context, Poll};
+    use tokio::io::unix::AsyncFd;
+
+    /// An async stream of events.
+    ///
+    /// This can be used through the [`futures::Stream`](Stream) implementation, or through simply
+    /// calling [`stream.next_event().await?`](Self::next_event).
+    pub struct EventStream {
+        device: AsyncFd<Device>,
+        events: VecDeque<InputEvent>,
+    }
+    impl Unpin for EventStream {}
+
+    impl EventStream {
+        pub(crate) fn new(device: Device) -> io::Result<Self> {
+            use nix::fcntl;
+            fcntl::fcntl(device.as_raw_fd(), fcntl::F_SETFL(fcntl::OFlag::O_NONBLOCK))
+                .map_err(nix_err)?;
+            let device = AsyncFd::new(device)?;
+            Ok(Self {
+                device,
+                events: VecDeque::new(),
+            })
+        }
+
+        /// Returns a reference to the underlying device
+        pub fn device(&self) -> &Device {
+            self.device.get_ref()
+        }
+
+        /// Try to wait for the next event in this stream. Any errors are likely to be fatal, i.e.
+        /// any calls afterwards will likely error as well.
+        pub async fn next_event(&mut self) -> io::Result<InputEvent> {
```

Comment:
Do we need to expose this or poll_event outside the crate?

Reply:
Yeah, I think this would be useful to have - I think I have to doc this better, but like - prefer calling `loop { next_event()? }`, since it retains error info, or else if you need a stream use stream combinators or whatever, or else if you're in a situation where you're implementing a future, use `poll_event`.